x-pack/plugins/timelines/common/utils/field_formatters.ts

```diff
@@ -51,117 +57,140 @@ export const isRuleParametersFieldOrSubfield = (field: string, prependField?: st
 export const isThreatEnrichmentFieldOrSubfield = (field: string, prependField?: string) =>
   prependField?.includes(ENRICHMENT_DESTINATION_PATH) || field === ENRICHMENT_DESTINATION_PATH;
 
+// Helper functions
+const createFieldItem = (
+  fieldCategory: string,
+  field: string,
+  values: string[],
+  isObjectArray: boolean
+): TimelineEventsDetailsItem => ({
+  category: fieldCategory,
+  field,
+  values,
+  originalValue: values,
+  isObjectArray,
+});
+
+const processGeoField = (
+  field: string,
+  item: unknown[],
+  fieldCategory: string
+): TimelineEventsDetailsItem => {
+  const formattedLocation = formatGeoLocation(item);
+  return createFieldItem(fieldCategory, field, formattedLocation, true);
+};
+
+const processSimpleField = (
+  dotField: string,
+  strArr: string[],
+  isObjectArray: boolean,
+  fieldCategory: string
+): TimelineEventsDetailsItem => createFieldItem(fieldCategory, dotField, strArr, isObjectArray);
+
+const processNestedFields = (
+  item: unknown,
+  dotField: string,
+  fieldCategory: string,
+  prependDotField: boolean
+): TimelineEventsDetailsItem[] => {
+  if (Array.isArray(item)) {
+    return item.flatMap((curr) =>
+      getDataFromFieldsHits(curr as Fields, prependDotField ? dotField : undefined, fieldCategory)
+    );
+  }
+
+  return getDataFromFieldsHits(
+    item as Fields,
+    prependDotField ? dotField : undefined,
+    fieldCategory
+  );
+};
+
+type DisjointFieldNames = 'ecs.version' | 'event.action' | 'event.kind' | 'event.original';
+
+// Memoized field maps
+const fieldMaps: EcsFieldMap &
+  Omit<TechnicalRuleFieldMap, DisjointFieldNames> &
+  ExperimentalRuleFieldMap = {
+  ...technicalRuleFieldMap,
+  ...ecsFieldMap,
+  ...legacyExperimentalFieldMap,
+};
+
 export const getDataFromFieldsHits = (
   fields: Fields,
   prependField?: string,
   prependFieldCategory?: string
-): TimelineEventsDetailsItem[] =>
-  Object.keys(fields).reduce<TimelineEventsDetailsItem[]>((accumulator, field) => {
+): TimelineEventsDetailsItem[] => {
+  const resultMap = new Map<string, TimelineEventsDetailsItem>();
+
+  for (const field of Object.keys(fields)) {
```

Comment:
Depending on the size of fields here, you might want to consider turning this into a good 'ol for-loop: https://jsbench.me/bum3sepcdq/1.

Reply:
done